### ISSUE TITLE
getPoolObj from macro parameters (sardana elements) randomly raises exceptions (SF#85)

### ISSUE THREAD

**Issue by sf-migrator-bot** [sardana-org/sardana#57]
Macro code randomly raises exceptions when trying to retrieve a Pool object from a Sardana element (passed to the macro code as a parameter e.g. a sardana moveable).

This is the trace:

```
SardanaTP.W003 DEBUG    2014-02-06 09:34:49,833 BL04/DOOR/01.MacroExecutor: [ENDEX] (MacroServerException) runMacro ipap_reset_motor
SardanaTP.W003 DEBUG    2014-02-06 09:34:49,835 BL04/DOOR/01: Traceback (most recent call last):
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/msmacromanager.py", line 1131, in runMacro
    for step in macro_obj.exec_():
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/macro.py", line 2143, in exec_
    res = self.run(*self._in_pars)
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/macro.py", line 2310, in run
    return self._function(self, *args)
  File "/homelocal/sicilia/lib/python/site-packages/macroserver/macros/icepap.py", line 361, in ipap_reset_motor
    pool_obj = motor.getPoolObj()
  File "/homelocal/sicilia/lib/python/site-packages/taurus/core/tango/sardana/pool.py", line 377, in getPoolObj
    return self._pool_obj
  File "/homelocal/sicilia/lib/python/site-packages/taurus/core/taurusdevice.py", line 77, in __getattr__
    return getattr(self._deviceObj, name)
  File "/homelocal/sicilia/lib/python/site-packages/PyTango/device_proxy.py", line 110, in __DeviceProxy__getattr
    raise AttributeError, name
AttributeError: _pool_obj
```

Reported by: reszelaz ( http://sf.net/u/zreszela )

Original Ticket: [sardana/tickets/85](https://sourceforge.net/p/sardana/tickets/85)

**Comment by reszelaz** [comment]
The same problem can be 100% reproduced with starting Pool and MacroServer, running a continuous scan and then running the `defmeas` macro:

```
Door_zreszela_1 [3]: ascanct mot02 0 10 10 1
ScanDir is not defined. This operation will not be stored persistently. Use Use "expconf" (or "senv ScanDir <abs directory>") to enable it
Scan #1678 started at Fri Jul 13 14:53:31 2018. It will take at least 0:00:00
Motor positions and relative timestamp (dt) columns contains theoretical values
   0         0         1         2         0.5       0.1   
   1         1         1         2         0.5       1.1   
   2         2         1         2         0.5       2.1   
   3         3         1         2         0.5       3.1   
   4         4         1         2         0.5       4.1   
   5         5         1         2         0.5       5.1   
   6         6         1         2         0.5       6.1   
   7         7         1         2         0.5       7.1   
   8         8         1         2         0.5       8.1   
   9         9         1         2         0.5       9.1   
   10        10        1         2         0.5       10.1  
Correcting overshoot...
Scan #1678 ended at Fri Jul 13 14:53:44 2018, taking 0:00:12.518271. Dead time 100.0% (motion dead time 100.0%)

Door_zreszela_1 [4]: defmeas cs-test2 ct01 ct12
An error occurred while running Macro 'defmeas(cs-test2, ['ct01', 'ct12']) -> cb4d5922-869b-11e8-be54-f8b156a48f79':
_pool_obj

Door_zreszela_1 [5]: www
Traceback (most recent call last):
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/msmacromanager.py", line 1462, in runMacro
    for step in macro_obj.exec_():
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/macro.py", line 2309, in exec_
    res = self.run(*self._in_pars)
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/macros/expert.py", line 89, in run
    pool = channel0.getPoolObj()
  File "/home/zreszela/workspace/sardana/src/sardana/taurus/core/tango/sardana/pool.py", line 418, in getPoolObj
    return self._pool_obj
  File "/home/zreszela/workspace/taurus/lib/taurus/core/tango/tangodevice.py", line 75, in __getattr__
    return getattr(self._deviceObj, name)
  File "/usr/lib/python2.7/dist-packages/tango/device_proxy.py", line 241, in __DeviceProxy__getattr
    raise AttributeError(name)
AttributeError: _pool_obj
```

**Comment by reszelaz** [comment]
This can be even simplified more to the `uct` and `defmeas`:

```
Door_zreszela_1 [1]: uct
      ct01       ct02  ioveri001
    1.0000     2.0000     0.5000

Door_zreszela_1 [2]: defmeas cs-test3 ct01 ct12
An error occurred while running Macro 'defmeas(cs-test3, ['ct01', 'ct12']) -> 8730cf7a-869c-11e8-8562-f8b156a48f79':
_pool_obj

Door_zreszela_1 [3]: www
Traceback (most recent call last):
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/msmacromanager.py", line 1462, in runMacro
    for step in macro_obj.exec_():
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/macro.py", line 2309, in exec_
    res = self.run(*self._in_pars)
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/macros/expert.py", line 89, in run
    pool = channel0.getPoolObj()
  File "/home/zreszela/workspace/sardana/src/sardana/taurus/core/tango/sardana/pool.py", line 418, in getPoolObj
    return self._pool_obj
  File "/home/zreszela/workspace/taurus/lib/taurus/core/tango/tangodevice.py", line 75, in __getattr__
    return getattr(self._deviceObj, name)
  File "/usr/lib/python2.7/dist-packages/tango/device_proxy.py", line 241, in __DeviceProxy__getattr
    raise AttributeError(name)
AttributeError: _pool_obj
```
But it must be `uct` and not normal `ct`.

**Comment by reszelaz** [comment]
The same happens for any scan. Start your MacroServer, run the `acsan` and later the `defmeas`.

```
oor_zreszela_1 [3]: ascan mot01 0 1 1 1
ScanDir is not defined. This operation will not be stored persistently. Use Use "expconf" (or "senv ScanDir <abs directory>") to enable it
Scan #1679 started at Fri Jul 13 17:18:57 2018. It will take at least 0:00:05.722854
 #Pt No    mot01      ct01      ct02    ioveri001     dt   
   0         0         1         2         0.5     2.86846 
   1         1         1         2         0.5     4.95948 
Scan #1679 ended at Fri Jul 13 17:19:03 2018, taking 0:00:05.998427. Dead time 66.7% (motion dead time 63.0%)

Door_zreszela_1 [4]: defmeas cs-test10 ct01
An error occurred while running Macro 'defmeas(cs-test10, ['ct01']) -> 18dd5138-86b0-11e8-9fc5-f8b156a48f79':
_pool_obj

Door_zreszela_1 [5]: www
Traceback (most recent call last):
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/msmacromanager.py", line 1462, in runMacro
    for step in macro_obj.exec_():
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/macro.py", line 2309, in exec_
    res = self.run(*self._in_pars)
  File "/home/zreszela/workspace/sardana/src/sardana/macroserver/macros/expert.py", line 89, in run
    pool = channel0.getPoolObj()
  File "/home/zreszela/workspace/sardana/src/sardana/taurus/core/tango/sardana/pool.py", line 418, in getPoolObj
    return self._pool_obj
  File "/home/zreszela/workspace/taurus/lib/taurus/core/tango/tangodevice.py", line 75, in __getattr__
    return getattr(self._deviceObj, name)
  File "/usr/lib/python2.7/dist-packages/tango/device_proxy.py", line 241, in __DeviceProxy__getattr
    raise AttributeError(name)
AttributeError: _pool_obj
```

**Comment by reszelaz** [comment]
I already found the reason for this. The problem happens if in your macro execution history the first time you access to a Pool element you do it directly using Taurus e.g.
```python
import taurus
channel = taurus.Device("ct01")
```
In this case the Tango factory gets populated with the correct extension device, in this case: `sardana.taurus.tango.sardana.pool.CTExpChannel` but this device misses some extra attributes which are filled when the same Pool element would be accessed from the [`Pool.getObject`](https://github.com/sardana-org/sardana/blob/debef451a383176985d9dd15bc89930ca69c3f60/src/sardana/taurus/core/tango/sardana/pool.py#L1932). These attributes are: `_pool_obj` and `_pool_data`.

The previosly listed examples of macrso: scans, `uct` are affected, and they may be quite a lot of users macros doing things in this way.

**Comment by guifrecuni** [comment]
Muy bien Zibi!

Gracias.

**Comment by reszelaz** [comment]
Solved in #865
